Initial Draft for ADT (#51)
Codecademy/ugc: content/goku-kun/introduction-to-adts-in-javascript.md

getHead(); // Returns back head node
getSize(); // Return the current size of the linked-list
isEmpty(); // Returns true if the linked-list is empty

// Insert and replace operations
insertBeginning(element); // Inserts new element at the beginning of the linked-list
insertEnd(element); // Inserts new element at the end of the linked-list
insertAtPosition(element, index); // Inserts new element at the given positional index
replaceAtPosition(element, index); // Replaces the element at give index with the new element

// Delete operations
deleteBeginning(); // Removes the first element and returns its element
deleteEnd(); // Removes the last element and returns its element
deleteAtPosition(index); // Removes node from the given positional index and returns its element

// Traverse, sort, and search operations
traverse(); // Goes through all the elements once in the linked list and prints them
search(element); // Searches given element and returns true if the element is found in linked-list
sort(order); // Sorts the linked-list in the given order (ascending/descending)
retrieve(index); // Returns the element stored at the given index location
```

## Stack ADTs

> **Stacks** are linear data structures in which data is entered and removed from _only a single point_. This point is called the **top** of the stack. It follows the last-in, first-out (L I F O) format for storing and discarding data. This means that the last element added to the top of the stack is the first element that will be removed from the stack. There is no other way to access other elements in the stack but the element that is at the top of the stack.

Reviewer: ```suggestion
**Stacks** are linear data structures in which data is entered and removed from _only a single point_. This point is called the **top** of the stack. It follows the last-in, first-out (LIFO) format for storing and discarding data. This means that the last element added to the top of the stack is the first element that will be removed from the stack. There is no other way to access other elements in the stack but the element that is at the top of the stack.
```

> ![Stack ADT](https://raw.githubusercontent.com/Codecademy/ugc/main/content/goku-kun/stack-adt.png)

Reviewer: ```suggestion
![Stack A D T](https://raw.githubusercontent.com/Codecademy/ugc/main/content/goku-kun/stack-adt.png)
```

A stack ADT supports the following operations:

```js
push(element); // Inserts a new element at the top of the stack
pop(); // Removes the element stored at the top of the stack and returns it
peek(); // Returns the top element without removing it from the stack
```

## Queue ADTs

**Queues** are linear data structures, in which data is inserted from one end and removed from the other end. The place where the data is inserted in the queue is called the **rear** end of the queue and this insertion operation is called **enqueue**. Data can be removed from the **front** of the queue and this deletion operation is called **dequeue**. It follows the first-in, first-out (FIFO) configuration for storing and removing data. This means that the data that was first to enter the queue is also first to leave the queue.

> ![Queue ADT](https://raw.githubusercontent.com/Codecademy/ugc/main/content/goku-kun/queue-adt.png)

Reviewer: ```suggestion
![Queue A D T](https://raw.githubusercontent.com/Codecademy/ugc/main/content/goku-kun/queue-adt.png)
```

## Introduction

> An **Abstract Data Type** (ADT), as the name suggests, is an abstract understanding of a data structure. An ADT is defined through its behavior and characteristics, particularly in terms of what data can be stored into it, the operations that can be performed on this data, and the behavior of these operations. For example, stacks and queues can be internally implemented using linked-lists made up of nodes, lists, or arrays. However, the primary function of a stack is to be a last in, first out (LIFO) data structure and the primary function of a queue is to be a first in, first out (FIFO) data structure. The behavior, from the point of the user, remains intact, regardless of the internal implementation either using linked-lists, arrays, or lists. If the user was interacting with a stack, the user will simply worry about pushing data onto the stack or popping data off the stack. The user won't need to have the knowledge of how that stack is working internally.

Reviewer: Thanks for bringing this up! This differentiation may alter across other languages that aren't JavaScript. So I'm thinking either we can just refer to arrays as "arrays" or if we want to keep "arrays/lists," we can mention that in JavaScript, the term list may be used synonymously with array?

> An **Abstract Data Type** (ADT), as the name suggests, is an abstract understanding of a data structure. An ADT is defined through its behavior and characteristics, particularly in terms of what data can be stored into it, the operations that can be performed on this data, and the behavior of these operations. For example, stacks and queues can be internally implemented using linked-lists made up of nodes, lists, or arrays. However, the primary function of a stack is to be a last in, first out (LIFO) data structure and the primary function of a queue is to be a first in, first out (FIFO) data structure. The behavior, from the point of the user, remains intact, regardless of the internal implementation either using linked-lists, arrays, or lists. If the user was interacting with a stack, the user will simply worry about pushing data onto the stack or popping data off the stack. The user won't need to have the knowledge of how that stack is working internally.

Author: I think since we've specifically said that the article is about JavaScript, we should remove the mention of lists altogether.

search(element); // Searches given element and returns true if the element is found in linked-list
sort(order); // Sorts the linked-list in the given order (ascending/descending)
retrieve(index); // Returns the element stored at the given index location
```

## Stack ADTs

**Stacks** are linear data structures in which data is entered and removed from _only a single point_. This point is called the **top** of the stack. It follows the last-in, first-out (L I F O) format for storing and discarding data. This means that the last element added to the top of the stack is the first element that will be removed from the stack. There is no other way to access other elements in the stack but the element that is at the top of the stack.

![Stack ADT](https://raw.githubusercontent.com/Codecademy/ugc/main/content/goku-kun/stack-adt.png)

A stack ADT supports the following operations:

```js
push(element); // Inserts a new element at the top of the stack
pop(); // Removes the element stored at the top of the stack and returns it
peek(); // Returns the top element without removing it from the stack
```

## Queue ADTs

> **Queues** are linear data structures, in which data is inserted from one end and removed from the other end. The place where the data is inserted in the queue is called the **rear** end of the queue and this insertion operation is called **enqueue**. Data can be removed from the **front** of the queue and this deletion operation is called **dequeue**. It follows the first-in, first-out (FIFO) configuration for storing and removing data. This means that the data that was first to enter the queue is also first to leave the queue.

Reviewer 2: I originally suggested this idea because I'd recently come across [this source](https://accessiblewebsiteservices.com/accessible-images-how-to-handle-alt-text-acronyms/) that describes delimiting acronyms with spaces in _alt text_ in images.

This isn't a huge requirement by any means, but something a11y-related to consider with regards to screen-readers that may read image alt text. For line 95, I would recommend we revert back to "LIFO" and then for lines like 97 and 111 (inside the brackets) we could add the spaces?

> **Queues** are linear data structures, in which data is inserted from one end and removed from the other end. The place where the data is inserted in the queue is called the **rear** end of the queue and this insertion operation is called **enqueue**. Data can be removed from the **front** of the queue and this deletion operation is called **dequeue**. It follows the first-in, first-out (FIFO) configuration for storing and removing data. This means that the data that was first to enter the queue is also first to leave the queue.

Author: I got confused. You were talking about alt text when in my earlier question, I was talking about the text inside the image itself. Haha. But sure. We could implement this.

> **Queues** are linear data structures, in which data is inserted from one end and removed from the other end. The place where the data is inserted in the queue is called the **rear** end of the queue and this insertion operation is called **enqueue**. Data can be removed from the **front** of the queue and this deletion operation is called **dequeue**. It follows the first-in, first-out (FIFO) configuration for storing and removing data. This means that the data that was first to enter the queue is also first to leave the queue.

Reviewer: whoops! Sorry about that. Thanks for interjecting @Dusch4593 to explain what was going on behind the scenes 😆 Yeah, originally we were thinking about accessibility for users, but the speech function works fine without the spacing. I'll add in the comments for what Dusch recommended for the in-line images
